#**TUGAS KELOMPOK PRAKTIKUM PBF RB**

##**Kelompok 3 Praktikum PBF RB**
Anggota Kelompok :
1. Deva Anjani Khayyuninafsyah (122450014)
2. Nabiilah Putri Karnaia (122450029)
3. Chevando Daffa Pramanda (122450095)
4. Novelia Adinda (122450104)
5. Rafly Prabu Darmawan (122450140)

##**Nomor 1**
Buatlah fungsi komposisi dengan nilai

f(x) = 4^x, g(x) = x^x

h(x) = 3 + x

dengan pemrograman prosedural tiga fungsi dan pemrograman fungsional dengan lambda.

In [125]:
# Pemrograman Prosedural dengan 3 Fungsi

import time

def f(x):
    return 4 ** x

def g(x):
    return x ** x

def h(x):
    return 3 + x

start_time = time.time()

# Contoh Pengaplikasian
x = 0.5
fungsi_komposisi = h(g(f(x)))

end_time = time.time()

elapsed_time = end_time - start_time

hasil = fungsi_komposisi
print("Hasil :", hasil)
print(f"Waktu : {elapsed_time:.9f} detik.")

Hasil : 7.0
Waktu : 0.000105858 detik.


In [126]:
# Pemrograman Fungsional dengan Lambda

import time

f = lambda x: 4**x
g = lambda x: x**x
h = lambda x: 3 + x

fungsi_komposisi = lambda x: h(g(f(x)))

start_time = time.time()

# Contoh Pengaplikasian
x = 0.5
fungsi_komposisi = h(g(f(x)))

end_time = time.time()

elapsed_time = end_time - start_time

print("Hasil :", fungsi_komposisi)
print(f"Waktu : {elapsed_time:.9f} detik.")

Hasil : 7.0
Waktu : 0.000082254 detik.


##**Nomor 2**
Buatlah pemrograman berbasis objek pada soal no. 1.

In [127]:
import time

class FungsiKomposisi:
  def __init__(self, formula):
    self.formula = formula

  def __call__(self, x):
    return eval(self.formula, {"x": x})

f = FungsiKomposisi("4**x")
g = FungsiKomposisi("x**x")
h = FungsiKomposisi("3 + x")

start_time = time.time()

# Contoh Pengaplikasian
x = 0.5
fungsi_komposisi = h(g(f(x)))

end_time = time.time()

elapsed_time = end_time - start_time

print("Hasil :", fungsi_komposisi)
print(f"Waktu : {elapsed_time:.9f} detik.")

Hasil : 7.0
Waktu : 0.000175953 detik.


##**Nomor 3**
Berikan analisis komparasi soal 1 dan 2.

Jawab:

Dari kode pemrograman yang telah dibuat pada nomor 1 dan 2, didapatkan:
1. Waktu Eksekusi Pemrograman Prosedural = 0.000105858 detik.
2. Waktu Eksekusi Pemrograman Fungsional = 0.000082254 detik.
3. Waktu Eksekusi OOP = 0.000175953 detik.

Sudah jelas dapat disimpulkan bahwa pemrograman fungsional dengan lambda berjalan lebih cepat dibandingkan kedua pemrograman lainnya.

Jika diperhatikan dengan lebih seksama, melalui modul 'time' ini dapat diketahui bahwa waktu eksekusi antara pemrograman prosedural, lambda, dan OOP memiliki nilai yang tidak jauh berbeda. Pada dasarnya, ketika program dijalankan, nilai waktu eksekusi dari ketika pemrograman tersebut akan menghasilkan nilai yang berbeda-beda.

Akan tetapi, dengan selisih perbedaan yang ada dapat diketahui pula bahwa, sebagian besar, pemrograman fungsional dengan lambda memiliki waktu eksekusi yang lebih cepat dibandingkan dengan pemrograman prosedural dan OOP.

##**Nomor 4**
Buatlah satu penyelesaian luas segitiga dengan menggunakan pemrograman berorientasi objek dan pemrograman berbasis fungsi. Uji dengan membandingkan mana yang tercepat komputasinya.

In [129]:
# Pemrograman Berorientasi Objek (PBO)

import time

class Segitiga:
    def __init__(self, alas, tinggi):
        self.alas = alas
        self.tinggi = tinggi

    def hitung_luas(self):
        return 0.5 * self.alas * self.tinggi

start_time = time.time()

# Contoh Pengaplikasian
alas = 18
tinggi = 9
segitiga = Segitiga(alas, tinggi)
luas = segitiga.hitung_luas()

end_time = time.time()

elapsed_time = end_time - start_time

print(f"Luas segitiga: {luas1} satuan luas.")
print(f"Waktu: {elapsed_time:.9f} detik.")

Luas segitiga: 81.0 satuan luas.
Waktu: 0.000207424 detik.


In [128]:
# Pemrograman Berorientasi Fungsi (PBF)

import time

def luas_segitiga(alas, tinggi):
    return 0.5 * alas * tinggi

start_time = time.time()

# Contoh Pengaplikasian
alas = 12
tinggi = 6
luas = luas_segitiga(alas, tinggi)

end_time = time.time()

elapsed_time = end_time - start_time

print(f"Luas segitiga: {luas2} satuan luas.")
print(f"Waktu: {elapsed_time:.9f} detik.")

Luas segitiga: 36.0 satuan luas.
Waktu: 0.000165939 detik.


Dari kedua kode pemrograman di atas, didapatkan:
1. Waktu Eksekusi Pemrograman Berorientasi Objek = 0.000207424 detik.
2. Waktu Eksekusi Pemrograman Berorientasi Fungsi = 0.000165939 detik.

Sudah jelas dapat disimpulkan bahwa pemrograman berbasis fungsi memiliki waktu eksekusi yang lebih cepat daripada pemrograman berorientasi objek.

##**Nomor 5**

Diberikan Algoritma Metode Bisection berikut:
1. Start
2. Define function f(x)
3. Choose initial guesses x0 and x1 such that f(x0)f(x1) < 0
4. Choose pre-specified tolarabe error e
5. Calculate new approximated root as x2 = (x0 + x1)/2
6. Calculate f(x0)f(x2)\
    a. if f(x0)f(x2) < 0 then x0 = x0 and x1 = x2\
    b. if f(x0)f(x2) > 0 then x0 = x2 and x1 = x1\
    c. if f(x0)f(x2) = 0 then go to (8)
7. if |f(x2)| > e then go to (5) otherwise go to (8)
8. Display x2 as root
9. Stop

Buatlah prosedural programming dari persoalan di atas dan ubahlah dengan functional programming. Lakukan dengan batas toleransi 10^-5. Dengan fungsi yang diberikan sebagai berikut: x - 2^x
dan nilai a dan b adalah di antara 0 dan 1 (nilai antara tidak boleh 0 dan 1).

In [153]:
# Pemrograman Prosedural

def f(x):
    return x - 2 ** x

def biseksi(f, x0, x1, tol):
    if f(x0) * f(x1) > 0:
      print("Tidak ada akar yang berada di antara x0 dan x1.")
      return None
    x2 = (x0 + x1) / 2
    while abs(x2) < tol:
        if  f(x0) * f(x2) < 0:
            x1 = x2
        elif f(x0) * f(x2) > 0:
            x0 = x2
        else:
          return x2
    return x2

print(biseksi(f, 0.25, 0.9, 0.00001))

Tidak ada akar yang berada di antara x0 dan x1.
None


In [164]:
# Pemrograman Fungsional

biseksi = (lambda f, x0, x1, tol:
    (lambda x0, x1:
        (lambda x2:
            x2 if abs(f(x2)) < tol else
            x0 if f(x0) * f(x2) < 0 else
            x1 if f(x0) * f(x2) > 0 else
            x2)(
            (x0 + x1) / 2
        )
    )(x0, x1) if f(x0) * f(x1) < 0 else (
        print("Tidak ada akar yang berada di antara x0 dan x1.") or None
    )
)

print(biseksi(lambda x: x - 2 ** x, 0.25, 0.9, 0.00001))

Tidak ada akar yang berada di antara x0 dan x1.
None
